src/input/keyboard/modifiers_state.rs

```diff
@@ -48,6 +48,75 @@ impl ModifiersState {
         self.iso_level5_shift = state.mod_name_is_active(&xkb::MOD_NAME_MOD3, xkb::STATE_MODS_EFFECTIVE);
         self.serialized = serialize_modifiers(state);
     }
+
+    /// Serialize modifier state back to be sent to xkb.
+    pub fn serialize_back(&self, state: &xkb::State) -> SerializedMods {
+        let keymap = state.get_keymap();
+
+        let mut locked: u32 = 0;
+        let mut depressed: u32 = 0;
+
+        if self.caps_lock {
+            let index = keymap.mod_get_index(&xkb::MOD_NAME_CAPS);
+            assert!(
+                index != xkb::MOD_INVALID,
+                "capslock is invalid modifier in keymap"
+            );
+            locked |= 1 << index;
+        }
+
+        if self.num_lock {
+            let index = keymap.mod_get_index(&xkb::MOD_NAME_NUM);
+            assert!(index != xkb::MOD_INVALID, "numlock is invalid modifier in keymap");
```

Comment:
Are these asserts guaranteed to pass even for weird custom keymaps? If not then I'd convert them to `if`s?

Reply:
I don't know tbh.

I'm not a fan of silently swallowing explicit requests by default, but I'm happy to remove the asserts if wanted.

I'll do more research and try do dig up relevant parts of the xkb spec.

Reply:
You *never* want to crash the compositor from something user-configurable (an xkb keymap)